src/haven/instrument/xray_source.py

```diff
@@ -1,32 +1,76 @@
 import logging
-import warnings
+from enum import IntEnum
 
-from apstools.devices.aps_undulator import ApsUndulator
-
-from .._iconfig import load_config
-from .device import make_device
+from ophyd import Component as Cpt
+from ophyd import DerivedSignal, Device, EpicsSignal, EpicsSignalRO, PVPositioner
 
 log = logging.getLogger(__name__)
 
 
-def load_xray_source(config=None):
-    if config is None:
-        config = load_config()
-    # Determine the X-ray source type (undulator vs bending magnet)
-
-    try:
-        data = config["xray_source"]
-    except KeyError:
-        warnings.warn("No X-ray source configured")
-    else:
-        if data["type"] == "undulator":
-            device = make_device(
-                ApsUndulator,
-                prefix=data["prefix"],
-                name="undulator",
-                labels={"xray_sources"},
-            )
-            return device
+class DoneStatus(IntEnum):
+    MOVING = 0
+    DONE = 1
+
+
+class BusyStatus(IntEnum):
+    DONE = 0
+    BUSY = 1
+
+
+class MotorDriveStatus(IntEnum):
+    NOT_READY = 0
+    READY_TO_MOVE = 1
+
+
+class UndulatorPositioner(PVPositioner):
+    setpoint = Cpt(EpicsSignal, "SetC.VAL")
+    readback = Cpt(EpicsSignalRO, "M.VAL")
+
+    actuate = Cpt(DerivedSignal, derived_from="parent.start_button", kind="omitted")
+    stop_signal = Cpt(DerivedSignal, derived_from="parent.stop_button", kind="omitted")
+    done = Cpt(DerivedSignal, derived_from="parent.done", kind="omitted")
+    done_value = DoneStatus.DONE
+
+
+class PlanarUndulator(Device):
```

Comment:
Is "PlanarUndulator" always "undulator"?

Reply:
Yes, 25-ID only has planar undulators. but other APS beamlines have other kinds.

Reply:
Here are the kinds: https://github.com/BCDA-APS/apstools/issues/973